samples/template/src/sidewalk.c

```diff
@@ -0,0 +1,353 @@
+	if (e) {
+		LOG_ERR("pal init err %d", (int)e);
+	}
+	e = sid_init(&sm->sid->config, &sm->sid->handle);
+	if (e) {
+		LOG_ERR("sid init err %d", (int)e);
+	}
+	e = sid_start(sm->sid->handle, sm->sid->config.link_mask);
+	if (e) {
+		LOG_ERR("sid start err %d", (int)e);
+	}
+}
+
+static void state_init_run(void *o)
+{
+	sm_t *sm = (sm_t *)o;
+
+	switch (sm->event) {
+	case SID_EVENT_STATE_READY:
+		smf_set_state(SMF_CTX(sm), &sid_states[STATE_SIDEWALK_READY]);
+		break;
+	case SID_EVENT_STATE_NOT_READY:
+		smf_set_state(SMF_CTX(sm), &sid_states[STATE_SIDEWALK_NOT_READY]);
+		break;
+	case SID_EVENT_CONNECT:
+	case SID_EVENT_SEND_MSG:
+		LOG_INF("Operation not supported, Sidewalk not initialized");
+		break;
+	case SID_EVENT_SIDEWALK:
+	case SID_EVENT_STATE_ERROR:
+	case SID_EVENT_FACTORY_RESET:
+	case SID_EVENT_LINK_SWITCH:
+	case SID_EVENT_NORDIC_DFU:
+		/* handled in common state */
+		break;
+	}
+}
+
+static void state_not_ready_run(void *o)
+{
+	sm_t *sm = (sm_t *)o;
+
+	switch (sm->event) {
+	case SID_EVENT_STATE_READY:
+		smf_set_state(SMF_CTX(sm), &sid_states[STATE_SIDEWALK_READY]);
+		break;
+	case SID_EVENT_SEND_MSG:
+		send_msg_pending = true;
+	case SID_EVENT_CONNECT:
+		sid_error_t e = sid_ble_bcn_connection_request(sm->sid->handle, true);
+		if (e) {
+			LOG_ERR("conn req err %d", (int)e);
+		}
+		break;
+	case SID_EVENT_STATE_NOT_READY:
+		break;
+	case SID_EVENT_SIDEWALK:
+	case SID_EVENT_STATE_ERROR:
+	case SID_EVENT_FACTORY_RESET:
+	case SID_EVENT_LINK_SWITCH:
+	case SID_EVENT_NORDIC_DFU:
+		/* handled in common state */
+		break;
+	}
+}
+
+static void state_ready_entry(void *o)
+{
+	if (send_msg_pending) {
+		send_msg_pending = false;
+		sidewalk_event_send(SID_EVENT_SEND_MSG);
+	}
+}
+
+static void state_ready_run(void *o)
+{
+	sm_t *sm = (sm_t *)o;
+	sid_error_t e = SID_ERROR_NONE;
+
+	switch (sm->event) {
+	case SID_EVENT_STATE_NOT_READY:
+		smf_set_state(SMF_CTX(sm), &sid_states[STATE_SIDEWALK_NOT_READY]);
+		break;
+	case SID_EVENT_SEND_MSG:
+		e = sid_put_msg(sm->sid->handle, &send_msg_buffer.msg, &send_msg_buffer.desc);
+		if (e) {
+			LOG_ERR("sid send err %d", (int)e);
+		}
+		LOG_INF("sid send (type: %d, id: %u)", (int)send_msg_buffer.desc.type,
+			send_msg_buffer.desc.id);
+		break;
+	case SID_EVENT_STATE_READY:
+	case SID_EVENT_CONNECT:
+		break;
+	case SID_EVENT_SIDEWALK:
+	case SID_EVENT_STATE_ERROR:
+	case SID_EVENT_FACTORY_RESET:
+	case SID_EVENT_LINK_SWITCH:
+	case SID_EVENT_NORDIC_DFU:
+		/* handled in common state */
+		break;
+	}
+}
+
+static void state_dfu_entry(void *o)
+{
+	sm_t *sm = (sm_t *)o;
+
+	LOG_INF("Entering into DFU mode");
+	int err = -ENOTSUP;
+#if defined(CONFIG_SIDEWALK_DFU_SERVICE_BLE)
+	err = nordic_dfu_ble_start();
+#endif
+	if (err) {
+		LOG_ERR("dfu start err %d", err);
+		smf_set_state(SMF_CTX(sm), &sid_states[STATE_SIDEWALK_INIT]);
+		return;
+	}
+}
+
+static void state_dfu_run(void *o)
+{
+	sm_t *sm = (sm_t *)o;
+
+	switch (sm->event) {
+	case SID_EVENT_NORDIC_DFU:
```

Comment:
So if I'm correct, receiving DFU event when we are already in DFU will result in exiting this state ? 

I can imagine that someone accidentally double-clicks the button, this will emit two DFU events, and we exit the DFU state, is that desirable ? Maybe we should rely on the fact that when we enter DFU state, the user can always force reset to start the application without performing DFU, or wait for a timeout to exit the DFU

Reply:
well I missed it was on long press, It can stay, just wanted to confirm that this is the intention